frame/staking/src/tests.rs

```diff
@@ -3859,6 +3860,87 @@ fn on_finalize_weight_is_nonzero() {
 	})
 }
 
+#[test]
+fn test_rebag() {
+	use crate::{
+		testing_utils::create_stash_controller,
+		voter_bags::{Bag, Node},
+	};
+	use frame_benchmarking::{whitelisted_caller};
+	use frame_system::RawOrigin;
+
+	const USER_SEED: u32 = 999666;
```

Comment:
seems like your tests are a little bit too much like the benchmakr? For tests, we have different setups. You can see the list of initial accounts (all of which are u64) who have funds.

You should first add some assertions to the test `basic_setup_works` and make sure these intial accounts are placed in the correct bag. For this, you need to pass a mock list of bag thresholds to the config trait. This will probably unravel that you forgot to insert genesis stakers into the new voting storage items. 

Then you should add tests that test these scenarios: 

- adding/removing both nominator and validator alters proper voting storage items.
- change someone's stash, and then call rebag. You should document that calling `bond_extra` and `rebond` now might need a rebag afterwards and open an issue to explain it to Jaco. 
- ...

Reply:
I mean, the test as it stands right now very intentionally is as close to the benchmark as I could get it, because there was about a week when the benchmark failed and the test passed and I couldn't figure out why. (In part, it's very hard to debug print stuff within the benchmark context!) (The reason turned out to be that it's not safe to assume that `u32::MAX` is bigger than an arbitrary account balance.) Agree that, now that the benchmark works, I should remove this test and write some more focused tests such as those you suggested.

Reply:
- [x] simplify rebag test: 22bc615a25c1f76ac42c1060061b74f9ab644fda
- [x] add some assertions to `basic_setup_works`: 4092fa8d5
- [x] adding/removing nominator and validator alters counts: https://github.com/paritytech/substrate/pull/9081/commits/fb5431e1782c60a2c59d0f7fa3a5c752c73bed35
- [ ] ~~document that `bond_extra` and `rebond` might need a rebag afterwards, explain to Jaco~~ make `bond_extra` and `rebond` call `rebag` internally, transparently